File: lab1/symtab.c
/**********************************************************************/
/*Per Emilsson och Kenny Pettersson                                   */
/**********************************************************************/

/**********************************************************************/
/* lab 1 DVG C01 - Symbol Table OBJECT                                */
/**********************************************************************/

/**********************************************************************/
/* Include files                                                      */
/**********************************************************************/
#include <stdio.h>
#include <ctype.h>
#include <string.h>
#include <stdbool.h>

/**********************************************************************/
/* Other OBJECT's METHODS (IMPORTED)                                  */
/**********************************************************************/
#include "keytoktab.h"

/**********************************************************************/
/* OBJECT ATTRIBUTES FOR THIS OBJECT (C MODULE)                       */
/**********************************************************************/
#define TABSIZE 1024                   /* symbol table size           */
#define NAMELEN   20                   /* name length                 */

typedef char tname[NAMELEN];

static tname  name[TABSIZE];
static toktyp role[TABSIZE];
static toktyp type[TABSIZE];
static int    size[TABSIZE];
static int    addr[TABSIZE];

static int addrindex=0;
static bool initialized=false;

static int numrows=0;                  /* number of rows in the ST    */
static int startp =0;                  /* start position program in ST*/

/**********************************************************************/
/*  PRIVATE METHODS for this OBJECT  (using "static" in C)            */
/**********************************************************************/

/**********************************************************************/
/*  GET methods (one for each attribute)                              */
/**********************************************************************/
static char * get_name(int ftref)   { return name[ftref]; }
static toktyp get_role(int ftref)   { return role[ftref]; }
static toktyp get_type(int ftref)   { return type[ftref]; }
static int    get_size(int ftref)   { return size[ftref]; }
static int    get_addr(int ftref)   { return addr[ftref]; }

/**********************************************************************/
/*  SET methods (one for each attribute)                              */
/**********************************************************************/
static void set_name(int ftref, char * fpname) { strcpy(name[ftref],
                                                        fpname);}
static void set_role(int ftref, toktyp frole)  { role[ftref] = frole; }
static void set_type(int ftref, toktyp ftype)  { type[ftref] = ftype; }
static void set_size(int ftref, int    fsize)  { size[ftref] = fsize; }
static void set_addr(int ftref, int    faddr)  { addr[ftref] = faddr; }

/**********************************************************************/
/*  Add a row to the symbol table                                     */
/**********************************************************************/
static void addrow(char *fname, toktyp frole, toktyp ftype,
                   int fsize, int faddr)
{
    set_name(numrows, fname);
    set_role(numrows, frole);
    set_type(numrows, ftype);
    set_size(numrows, fsize);
    set_addr(numrows, faddr);
    numrows++;
}
/**********************************************************************/
/*  Initialise the symbol table                                       */
/**********************************************************************/
static void initst()
{
    addrow(tok2lex(predef),  typ, predef, 0, 0);
    addrow(tok2lex(undef),   typ, predef, 0, 0);
    addrow(tok2lex(error),   typ, predef, 0, 0);
    addrow(tok2lex(integer), typ, predef, 4, 0);
    addrow(tok2lex(boolean), typ, predef, 4, 0);
    addrow(tok2lex(real),    typ, predef, 8, 0);
    initialized = true;
}
/**********************************************************************/
/*  return a reference to the ST (index) if name found else nfound    */
/**********************************************************************/
static int get_ref(char * fpname)
{
    for(int i = 0; i < startp; i++){
        if(strcmp(name[i], fpname) == 0){
            return i;
        }
    } 
    return 0;
}

static void updatevar(int varsize, int varaddr, int index){
    size[index] = varsize;
    addr[index] = varaddr;
    addrindex += varsize;
    set_size(startp,get_size(startp)+varsize);
}

/**********************************************************************/
/*  PUBLIC METHODS for this OBJECT  (EXPORTED)                        */
/**********************************************************************/
/**********************************************************************/
/*  Display the symbol table                                          */
/**********************************************************************/
static void p_symrow(int ftref)
{
    printf("\n%-12s %4s %s %d %d", get_name(ftref), tok2lex(get_role(ftref)), tok2lex(get_type(ftref)), get_size(ftref), get_addr(ftref));
}

void p_symtab()
{
    if(numrows>0){
        for(int i = startp; i < numrows; i++){
			p_symrow(i);
        }
    }
	printf("\nSTATIC STORAGE REQUIRED is %d BYTES\n", get_size(startp));
}

/**********************************************************************/
/*  Add a program name to the symbol table                            */
/**********************************************************************/
void addp_name(char * fpname)
{
    if(!initialized){initst();}

    startp = numrows;
    addrow(fpname, program, program, 0, 0);
}

/**********************************************************************/
/*  Add a variable name to the symbol table                           */
/**********************************************************************/
void addv_name(char * fpname)
{
    addrow(fpname, var, var, 0, 0);
}

/**********************************************************************/
/*  Find a name in the the symbol table                               */
/*  return a Boolean (true, false) if the name is in the ST           */
/**********************************************************************/
int find_name(char * fpname)
{
    for(int i = 0; i < numrows; i++){
        if(strcmp(name[i], fpname) == 0){
            return 1;
        }
    } 
    return 0;
}

/**********************************************************************/
/*  Set the type of an id list in the symbol table                    */
/**********************************************************************/
void setv_type(toktyp ftype)
{
    if(numrows>0){
        for(int i = 0; i < numrows; i++){
            if(type[i] == var){
                type[i] = ftype;
                switch(ftype){
                    case(integer):
                        updatevar(get_size(get_ref("integer")), addrindex, i);
                        break;
                    case(real):
                        updatevar(get_size(get_ref("real")), addrindex, i);
                        break;
                    case(boolean):
                        updatevar(get_size(get_ref("boolean")), addrindex, i);
                        break;
					case (error):
						updatevar(get_size(get_ref("error")), addrindex, i);
						break;
					default:
						break;
                }
            }
        }
    }
}

/**********************************************************************/
/*  Get the type of a variable from the symbol table                  */
/**********************************************************************/
toktyp get_ntype(char * fpname)
{
    for(int i = 0; i < numrows; i++){
        if(strcmp(name[i], fpname) == 0){
            return type[i];
        }
    } 
    return 0;
}

/**********************************************************************/
/* End of code                                                        */
/**********************************************************************/

File: lab4/symtab.py
# **********************************************************************
# *Per Emilsson och Kenny Pettersson                                   *
# **********************************************************************

from dataclasses import dataclass
from typing import List, Optional

from keytoktab import *

@dataclass
class SymbolEntry:
    name: str
    role: toktyp
    type: toktyp
    size: int
    addr: int

class SymbolTable:
    def __init__(self):
        self.entries: List[SymbolEntry] = []
        self.initialized: bool = False
        self.program_index: Optional[int] = None #Optional is used to say that this will hold an int or None
        self.total_size: int = 0
    
    def init(self):
        if self.initialized:
            return
            
        # predefined types and error entries
        for t in (toktyp.PREDEF, toktyp.UNDEF, toktyp.ERROR,
                  toktyp.INTEGER, toktyp.BOOLEAN, toktyp.REAL):
            if t in (toktyp.INTEGER, toktyp.BOOLEAN):
                size = 4
            elif t == toktyp.REAL:
                size = 8
            else:
                size = 0
            entry = SymbolEntry(
                name=tok2lex(t),
                role=toktyp.TYPE,
                type=toktyp.PREDEF,
                size=size,
                addr=0
            )
            self.entries.append(entry) #add the entry to our entries List
        self.initialized = True
        
    def addp_name(self, name: str):
        self.init()

        self.program_index = len(self.entries)
        self.entries.append(SymbolEntry(
            name=name,
            role=toktyp.PROGRAM,
            type=toktyp.PROGRAM,
            size=0,
            addr=0
        ))
        self.total_size = 0
        
    def addv_name(self, name: str) -> int:
        numrows = len(self.entries)
        self.entries.append(SymbolEntry(
            name=name,
            role=toktyp.VAR,
            type=toktyp.VAR,
            size=0,
            addr=0
        ))
        return numrows
    
    def find_name(self, name: str) -> Optional[int]:
        for i, entry in enumerate(self.entries):
            if entry.name == name:
                return i
        return None
        
    def setv_type(self, vtype: toktyp):
        if self.program_index is None:
            return
        # Only assign type and address to variables not yet typed (entry.type == VAR)
        for entry in self.entries:
            if entry.role == toktyp.VAR and entry.type == toktyp.VAR:
                entry.type = vtype
                # lookup size of this type
                type_idx = self.find_name(tok2lex(vtype)) or 0
                entry.size = self.entries[type_idx].size
                entry.addr = self.total_size
                self.total_size += entry.size
        # update total program size
        self.entries[self.program_index].size = self.total_size
        
    def get_ntype(self, name: str) -> toktyp:
        idx = self.find_name(name)
        return self.entries[idx].type if idx is not None else toktyp.ERROR #return type or error if the name is not in the table
    
    def p_symtab(self):
        if self.program_index is not None:
            print(f"\n________________________________________________________")
            print(f" THE SYMBOL TABLE")
            print(f"________________________________________________________")
            print(f" NAME         ROLE TYPE SIZE ADDR")
            for entry in self.entries[self.program_index:]:
                print(f"{entry.name:<12} {tok2lex(entry.role):>4} {tok2lex(entry.type)} {entry.size} {entry.addr}")
            print(f"\nSTATIC STORAGE REQUIRED is {self.total_size}")